# Ensemble Learning

## Initial Imports

In [109]:
import warnings
warnings.filterwarnings('ignore')

In [110]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter
from sklearn.ensemble import RandomForestClassifier
from imblearn.ensemble import EasyEnsembleClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
%matplotlib inline

In [111]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [112]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)
df = pd.get_dummies(df, columns=["home_ownership", "verification_status","hardship_flag","debt_settlement_flag","issue_d","pymnt_plan","application_type","initial_list_status","next_pymnt_d"])
# Preview the data
df.head()

loan_amnt  int_rate  installment  annual_inc loan_status    dti  \
0    10500.0    0.1719       375.35     66000.0    low_risk  27.24   
1    25000.0    0.2000       929.09    105000.0    low_risk  20.23   
2    20000.0    0.2000       529.88     56000.0    low_risk  24.26   
3    10000.0    0.1640       353.55     92000.0    low_risk  31.44   
4    22000.0    0.1474       520.39     52000.0    low_risk  18.76   

   delinq_2yrs  inq_last_6mths  open_acc  pub_rec  ...  issue_d_Feb-2019  \
0          0.0             0.0       8.0      0.0  ...                 0   
1          0.0             0.0      17.0      1.0  ...                 0   
2          0.0             0.0       8.0      0.0  ...                 0   
3          0.0             1.0      10.0      1.0  ...                 0   
4          0.0             1.0      14.0      0.0  ...                 0   

   issue_d_Jan-2019  issue_d_Mar-2019  pymnt_plan_n  \
0                 0                 1             1   
1                 0                 1             1   
2                 0                 1             1   
3                 0                 1             1   
4                 0                 1             1   

   application_type_Individual  application_type_Joint App  \
0                            1                           0   
1                            1                           0   
2                            1                           0   
3                            1                           0   
4                            1                           0   

   initial_list_status_f  initial_list_status_w  next_pymnt_d_Apr-2019  \
0                      0                      1                      0   
1                      0                      1                      0   
2                      0                      1                      0   
3                      0                      1                      0   
4                      0                      1                      0   

   next_pymnt_d_May-2019  
0                      1  
1                      1  
2                      1  
3                      1  
4                      1  

[5 rows x 96 columns]

## Split the Data into Training and Testing

In [113]:
# Create our features
x_cols = [i for i in df.columns if i not in ("loan_status")]
X = df[x_cols]
X
# Create our target
y = df['loan_status']
y.head()

0    low_risk
1    low_risk
2    low_risk
3    low_risk
4    low_risk
Name: loan_status, dtype: object

In [114]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  issue_d_Feb-2019  issue_d_Jan-2019  \
count   68817.000000  ...      68817.000000      68817.000000   
mean    17604.142828  ...          0.371696          0.451066   
std     21835.880400  ...          0.483261          0.497603   
min         0.000000  ...          0.000000          0.000000   
25%      6293.000000  ...          0.000000          0.000000   
50%     12068.000000  ...          0.000000          0.000000   
75%     21735.000000  ...          1.000000          1.000000   
max    587191.000000  ...          1.000000          1.000000   

       issue_d_Mar-2019  pymnt_plan_n  application_type_Individual  \
count      68817.000000       68817.0                 68817.000000   
mean           0.177238           1.0                     0.860340   
std            0.381873           0.0                     0.346637   
min            0.000000           1.0                     0.000000   
25%            0.000000           1.0                     1.000000   
50%            0.000000           1.0                     1.000000   
75%            0.000000           1.0                     1.000000   
max            1.000000           1.0                     1.000000   

       application_type_Joint App  initial_list_status_f  \
count                68817.000000           68817.000000   
mean                     0.139660               0.123879   
std                      0.346637               0.329446   
min                      0.000000               0.000000   
25%                      0.000000               0.000000   
50%                      0.000000               0.000000   
75%                      0.000000               0.000000   
max                      1.000000               1.000000   

       initial_list_status_w  next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  
count           68817.000000           68817.000000           68817.000000  
mean                0.876121               0.383161               0.616839  
std                 0.329446               0.486161               0.486161  
min                 0.000000               0.000000               0.000000  
25%                 1.000000               0.000000               0.000000  
50%                 1.000000               0.000000               1.000000  
75%                 1.000000               1.000000               1.000000  
max                 1.000000               1.000000               1.000000  

[8 rows x 95 columns]

In [115]:
# Check the balance of our target values
y.value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [116]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [117]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [118]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [119]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [120]:
# Resample the training data with the BalancedRandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=500, random_state=78)
rf_model = rf_model.fit(X_train_scaled, y_train)
y_pred = rf_model.predict(X_test_scaled)

In [121]:
# Calculated the balanced accuracy score
acc_score = accuracy_score(y_test, predictions)
print(acc_score)

0.9961057832025574


In [122]:
# Display the confusion matrix
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)

In [123]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.93      0.37      1.00      0.52      0.61      0.34       101
   low_risk       1.00      1.00      0.37      1.00      0.61      0.39     17104

avg / total       1.00      1.00      0.37      1.00      0.61      0.39     17205



In [131]:
# List the features sorted in descending order by feature importance
importances = rf_model.feature_importances_
sorted(zip(rf_model.feature_importances_, X.columns), reverse=True)
importances_df = pd.DataFrame(sorted(zip(rf_model.feature_importances_, X.columns), reverse=True))
importances_df.set_index(importances_df[1], inplace=True)
importances_df.drop(columns=1, inplace=True)
importances_df.rename(columns={0: 'Feature Importances'}, inplace=True)
importances_sorted = importances_df.sort_values(by='Feature Importances')
#importances_sorted.plot(kind='barh', color='lightgreen', title= 'Features Importances', legend=False)
importances_sorted.tail()

Feature Importances
1                                   
total_rec_prncp             0.058795
total_pymnt                 0.059640
total_pymnt_inv             0.066507
last_pymnt_amnt             0.069615
total_rec_int               0.071884

### Easy Ensemble Classifier

In [125]:
# Train the Classifier
ee_model = EasyEnsembleClassifier(n_estimators=500, random_state=78)
ee_model = ee_model.fit(X_train_scaled, y_train)
y_pred = ee_model.predict(X_test_scaled)

In [126]:
# Calculated the balanced accuracy score
acc_score = accuracy_score(y_test, predictions)
print(acc_score)

0.9961057832025574


In [127]:
# Display the confusion matrix
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)

In [128]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.09      0.92      0.95      0.17      0.93      0.87       101
   low_risk       1.00      0.95      0.92      0.97      0.93      0.87     17104

avg / total       0.99      0.95      0.92      0.97      0.93      0.87     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    Easy Essmbly: 0.9469921534437663, Random Forest: 0.9961057832025574

2. Which model had the best recall score?

    Easy Essembly: 0.95 Random Forest:  1.00 HOWEVER Random forest has very low recall score for high risk! which is what we are trying to flag ( 0.37 ) as opposed to easy (0.92 )

3. Which model had the best geometric mean score?

    Easy Essembly:0.93andom Forest Macro Average: 0.61

4. What are the top three features?
otal_rec_int, last_pymnt_amnt, total_pymnt_inv